### DISCUSSION TITLE
Full duplex traffic between two esp32s?

### DISCUSSION THREAD

**Discussion by albinotoni** [pschatzmann/arduino-audio-tools#361] [Q&A] [answered]
Hello,

I'm trying to figure out whether it would be possible to use two esp32s kind of as a 'telephone' or 'full duplex walkie-talkie' system? 
 
Could something like this work with arduino-audio?:

I2S microphone (adafruit MEMS breakout) -> (esp32) -> ip/wifi -> (esp32) -> DAC out 

..and the same the other way around with the same ports. Though I haven't even quite catch up yet could I use the same I2S-port for input and output simultaneously?

**Comment by pschatzmann** [comment, marked as answer]
Yes, I2S is duplex and you can read and write at the same time on the same port.
I would recommend to use a [Codec](https://github.com/pschatzmann/arduino-audio-tools/wiki/Encoding-and-Decoding-of-Audio) to minimize the volume of data.
The only potential issue that I see in doing this, is acoustic feedback..

**Reply by albinotoni** [reply]
Thank you for the quick answer!

The feedback is a problem. I was planning to mute or lower the mic input stream according to the output signal strength as I saw there is also this 'volume-stream'. Maybe a somekind of gating system also, dirty compander system or something and placement of the mic, we'll see. I'm not after a high fidelity on this one..

**Reply by dragoncoder047** [reply]
On the placement of the mic -- you could just wire your ADC and DAC up to a TRRS audio jack, and plug in a headset.

**Comment by albinotoni** [comment]
Hey again. I just until now got an i2s mic to continue on this.. 
I've manage to send prerecorded wav through ip (wifi) between esp32s and play it through analog out.
Also I've managed to copy i2s to csvstream on one esp32 and I think also between two esps (at least I got something on the plotter but it's quite hard to read). This was 32 bit samples. 16 bit didn't work. 
But still I haven't succeed to copy the i2s mic signal to the analog out on other esp. The best I've got is static noise. I tried to config the signal to 16 bit (since AnalogAudioStreams read function is using 16 bit ?) but then it got all mute. 
Should I decode the client-stream somehow? (not because of the reduced data volume, I'm not that far yet..) Or is the reason in the 32/16 bits? The microphone  Adafruit's i2s mems.
Surely this would be easier and probably better by using also i2s dac on the other end, but I don't have one at the moment.

Other question: When receiving wav over ip, if the sender stops sending, the receiver keeps looping whatever is left on its buffer. What's the correct method to check whether there's new data coming? ( (if client.connected()) and (if client) seem to check only if the wifi is connected?

All in all your libraries are very straightforwarded and thanks to many examples, I've got even this far! So nice!

**Reply by dragoncoder047** [reply]
It's `client.avaliable()` to check if the buffer still has data. (Docs: https://www.arduino.cc/reference/en/libraries/wifi/client.available/) Hope that helps.

As to why the mic doesn't seem to be working, I have absolutely no idea.

**Reply by albinotoni** [reply]
Ok, thanks for that! I really need to start looking closer to libraries I'm using!

**Reply by albinotoni** [reply]
Unfortunaltely it seems that at least

if (client.available()){
    copier.copy();
  }

doesn't help. If the sender stops sending (power off or something else), the receiver keeps looping the buffer. So the buffer never flushes, only overwrites or what? (I really don't know).

**Reply by pschatzmann** [reply]
I suggest that you have another look at the provided link. It shows how to determine and handle closed sessions

**Comment by pschatzmann** [comment]
If you send 32 data over the write and write it out via AnalogAudioStreams you need to set this in the config. By default 16 bit is proposed...
If you use 16 bit you need to analyse the signal on both channels (with a CSVStream): The best thing is to [record a sine wave](https://www.szynalski.com/tone-generator/) or to whistle: this way you can see easily on which channel you get a valid signal and which channel you need to filter out (or overwrite)...

**Reply by albinotoni** [reply]
Now I managed to transfer signal, but it's very weak and the samplerate is mismatched (it's set the same on sender and receiver and works if I transfer for example the alice-sample.. I think I'll try with an external i2s dac sometimes later, and maybe try a mic of different brand also..
